volatility3/framework/plugins/windows/mftscan.py

```diff
@@ -189,3 +190,158 @@ def run(self):
+        )
+
+        # get each of the individual Field Sets
+        mft_object = symbol_table + constants.BANG + "MFT_ENTRY"
+        header_object = symbol_table + constants.BANG + "ATTR_HEADER"
+        attribute_object = symbol_table + constants.BANG + "ATTRIBUTE"
+        fn_object = symbol_table + constants.BANG + "FILE_NAME_ENTRY"
+
+        # Scan the layer for Raw MFT records and parse the fields
+        for offset, _rule_name, _name, _value in layer.scan(
+            context=self.context, scanner=yarascan.YaraScanner(rules=rules)
+        ):
+            with contextlib.suppress(exceptions.PagedInvalidAddressException):
+                mft_record = self.context.object(
+                    mft_object, offset=offset, layer_name=layer.name
+                )
+                # We will update this on each pass in the next loop and use it as the new offset.
+                attr_base_offset = mft_record.FirstAttrOffset
+
+                attr_header = self.context.object(
+                    header_object,
+                    offset=offset + attr_base_offset,
+                    layer_name=layer.name,
+                )
+
+                # There is no field that has a count of Attributes
+                # Keep Attempting to read attributes until we get an invalid attr_header.AttrType
+                file_name = "N/A"
+                is_ads = False
+                    # The First $DATA Attr is the 'principal' file itself not the ADS
+                while attr_header.AttrType.is_valid_choice:
+                    # Offset past the headers to the attribute data
+                    attr_data_offset = (
+                        offset
+                        + attr_base_offset
+                        + self.context.symbol_space.get_type(
+                            attribute_object
+                        ).relative_child_offset("Attr_Data")
+                    )
+
+                    if attr_header.AttrType.lookup() == "FILE_NAME":
+                        attr_data = self.context.object(
+                            fn_object, offset=attr_data_offset, layer_name=layer.name
+                        )
+                        file_name = attr_data.get_full_name()
+                    
+
+                    # DATA Attribute (can be ADS or not)
+                    if attr_header.AttrType.lookup() == "DATA":
+                        if is_ads:
+                            if not attr_header.NonResidentFlag:
+                                # Resident files are the most interesting.
+                                if attr_header.NameLength > 0:
+                                    attr_name_offset = (
+                                        offset
+                                        + attr_base_offset
+                                        + attr_header.NameOffset
+                                    )
+                                    ads_name = self._context.layers[layer.name].read(
```

Comment:
Again, please cast this a `string` of some kind, rather than reimplementing the string decoding stuff again.

Reply:
No sure I understood, 
The ADS filename like the ADS content (unlike the $FILE_NAME attr) is not at a predictable offset, let me know if I've missed something and your insight on how I could do the string casting in a better way  😊